### DISCUSSION TITLE
distinct operator?

### DISCUSSION THREAD

**Discussion by hillin** [iter-tools/iter-tools#479] [Ideas]
I'm asking here first in case I overlooked the api docs again: is there a `distinct` operator? I.e. it only returns the first occurrence of each distinct element in an iterable.

e.g. `distinct([3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8])` returns `[3, 1, 4, 5, 9, 2, 6, 8]` .

**Comment by conartist6** [comment]
You're describing `Set`.
```js
equal(new Set([3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8]), [3, 1, 4, 5, 9, 2, 6, 8]); // true
```

In addition iter-tools offers [splitGroups](https://github.com/iter-tools/iter-tools/blob/v7.3.2/API.md#splitgroups) which can merge adjacent values.

**Reply by conartist6** [reply]
I admit that it would be quite possible to add `distinct` and `group` methods which use `Set` and `Map` internally, but so far I have avoided adding methods whose implementation has no iterative characteristics, i.e. the results cannot be computed lazily or partially.

**Reply by hillin** [reply]
`Set` is exact what I'm using right now, but it has a down side that if I need to filter out the duplicated elements in the middle of a pipeline, I have to break the pipeline apart.

Also it's not totally true that the `distinct` operator does not have *iterative characteristics*. With the `Set` approach you have to create a set out of the entire iterable, but ideally the `distinct` operator, although very likely uses a `Set` internally, builds the set progressively. I.e. only when an value is yielded by the upstream and fed to the `distinct` operator, its identity is checked against the Set and added if it's distinct. This makes a huge difference if the upstream produces a large amount of elements, even infinity of them, because the pipeline could terminate the iteration early by using operators such as `take`.

**Reply by conartist6** [reply]
You don't have to break the pipeline, you could just put an arrow in like `iterable => new Set(iterable)`. It kinda depends on personal preferences really.

I see what you're saying about the iterative aspect of `distinct`. In that way it's actually different than `group`, since `group` does definitely require you to consume the whole input so that you see all the values that may belong in a particular group.

I guess I'd say that means I'm willing to add `distinct`. Just out of curiosity, what do you need it for?

**Reply by hillin** [reply]
My current use cases can be fully covered by using a Set, I always consume the entire collection to find distinctive values. It's just that from a C# (LINQ) and rx background, distinct is the goto solution when I want to remove duplicated values, so I thought iter-tools should've already have it too. But then again I agree for now Set is all I need, and the in-pipeline Set construction looks promising (guess my mindset is a little bit limited thanks to using too much rx, where you can't pipe things as freely).

For use cases where this operator may really shine, imagine you have a huge song list and want to find the first ten artists to display. 

```
execPipe(
  songs,
  map(song => song.artist),
  distinct(),
  take(10)
)
```

**Reply by conartist6** [reply]
OK yeah. Here's that logic (more or less -- with 4 albums instead of 10 artists) in the real world (Spotify):

<img width="411" alt="Screen Shot 2022-06-09 at 11 29 22 AM" src="https://user-images.githubusercontent.com/540777/172885664-871bd181-8a1c-4524-ad4a-1638910cfa49.png">

**Reply by conartist6** [reply]
Do you want to build it or should I? I can probably do it faster, but I think it's better for the health of the project in the long run to get more people involved in the code.

**Reply by conartist6** [reply]
Should the method take an optional `same` argument like `equal` does?

**Reply by hillin** [reply]
I'd be glad to contribute, I'll see what I can do shortly.

I'm tempted to say it should be able to receive an equality comparer, but that means we can't use a Set internally(which does not support such comparer), and it's less performance-friendly (O(n) as opposed to O(logn)) in terms of checking existence.

A viable alternative is to provide another variant, `distinctBy`, which filters an iterable by certain criteria, specified with a transform function (e.g. `distinctBy(entity => entity.id)`). 

BTW I think there are some other APIs of iter-tools which can benefit from having a `-By` variant too. I always missed the `OrderBy` siblings in LINQ when I want to sort something in JS.

**Reply by conartist6** [reply]
yeah `distinct` and `distinctBy` is more like what I've done elsewhere in the library. In this case there would be no need to have two names necessarily since an iterable is distinguishable from a function.

**Reply by conartist6** [reply]
What I mean is that when a function has an optional argument the currying engine has to decide whether to do a partial application of the optional argument, or whether to execute the method and return a result. It assumes that anything that isn't iterable must be partial application. In places where the optional argument could actually be an iterable then I choose multiple names. That's why there are separate `split` and `splitOn` functions -- the value to split on could be an array.

**Reply by conartist6** [reply]
Also make sure to check out the `CONTRIBUTING` file. My code generator is not very elegant. I'm working on replacing it, but the work keeps getting tangled up in other projects (rewriting recast).

**Reply by hillin** [reply]
Am I also supposed to create an async version of this operator? I haven't touched iter-tools' async api yet.

**Reply by hillin** [reply]
I made a little attempt and it seems I shouldn't have tried on Windows, the differences of the newline (crlf) and path delimiter ('\\' not '/') has made quite a mess. Will try again later when I can access my Ubuntu machine.

**Reply by conartist6** [reply]
Oh darn, I'm sorry about that. How bad a mess? I've never tested the built tools on windows. Thee code generator should in theory be able to clean up any mess it makes, but if adding files made a mess I'd be very hesitant to find out what kind of problems untested code for deleting files could cause. I do my dev on OSX, CI is on Ubuntu, and  I've also verified that it's possible to develop under Windows using WSL2 Ubuntu/VSCode remote. My intention is for the completed build system to work fine on windows, but as mentioned it's not done.

**Reply by conartist6** [reply]
You'll need to create an async version of the method, yes, but the build system will help you do it. You can do `yarn create:method -b -m distinct` to generate the basic scaffolding you need, then edit `src/impls/distinct/$distinct.js`. To understand how the `$` works to help you build sync and async code at the same time look at the comments in [async.macro](https://github.com/iter-tools/iter-tools/blob/17452c9ad901094e9499e45f4bed9b844ce50977/generate/async.macro.cjs).

**Reply by hillin** [reply]
> How bad a mess?

Over a thousand files were affected, mostly due to the newline and path delimiter difference. A few tests didn't pass too. Anyways I'm getting my hands on the linux box now.

**Reply by hillin** [reply]
#480
